Desugar module var and accessor in refchecks/lazyvals

retronym opened the pull request:
Rather than leaving it until mixin.

The broader motivation is to simplify the mixin phase of the
compiler before we get rid of implementatation classes in
favour of using JDK8 default interface methods.

The current code in mixin is used for both lazy val and modules,
and puts the "slow path" code that uses the monitor into a
dedicated method (`moduleName$lzyCompute`). I tracked this
back to a3d4d17b77. I can't tell from that commit whether the
performance sensititivity was related to modules or lazy vals,
from the commit message I'd say the latter.

As the initialization code for a module is just a constructor call,
rather than an arbitraryly large chunk of code for a lazy initializer,
this commit opts to inline the `lzycompute` method.

During refchecks, mixin module accessors are added to classes, so
that mixed in and defined modules are translated uniformly.

Defer synthesis of the double checked locking idiom to the lazyvals
phase, which gets us a step closer to a unified translation of
modules and lazy vals.

I had to change the `atOwner` methods to to avoid using the
non-existent module class of a module accessor method as the
current owner. This fixes a latent bug. Without this change,
retypechecking of the module accessor method during erasure crashes
with an accessibility error selecting the module var.

Review by @adriaanm 

I'll add a results of bytecode diffing to the comments here in a day or so. I need to rig up your ASM utilities to do this cleanly. There will be changes to the names of some synthetic private methods because we used to lambda lift the module and the create the accessor in mixin, whereas now we create the "$lzycompute" method before lambda lift (in lazy vals).

Here's a preview of the changes in trees I'm looking at: https://gist.github.com/retronym/855857c1c1bf9452f1b8

retronym committed:
Desugar module var and accessor in refchecks/lazyvals

Rather than leaving it until mixin.

The broader motivation is to simplify the mixin phase of the
compiler before we get rid of implementatation classes in
favour of using JDK8 default interface methods.

The current code in mixin is used for both lazy val and modules,
and puts the "slow path" code that uses the monitor into a
dedicated method (`moduleName$lzyCompute`). I tracked this
back to a3d4d17b77. I can't tell from that commit whether the
performance sensititivity was related to modules or lazy vals,
from the commit message I'd say the latter.

As the initialization code for a module is just a constructor call,
rather than an arbitraryly large chunk of code for a lazy initializer,
this commit opts to inline the `lzycompute` method.

During refchecks, mixin module accessors are added to classes, so
that mixed in and defined modules are translated uniformly. Trait
owned modules get an accessor method with an empty body (that shares
the module symbol), but no module var.

Defer synthesis of the double checked locking idiom to the lazyvals
phase, which gets us a step closer to a unified translation of
modules and lazy vals.

I had to change the `atOwner` methods to to avoid using the
non-existent module class of a module accessor method as the
current owner. This fixes a latent bug. Without this change,
retypechecking of the module accessor method during erasure crashes
with an accessibility error selecting the module var.

In the process, I've tweaked a tree generation utility method
to wvoid synthesizing redundant blocks in module desugaring.

retronym commented:
I missed the review comments on this one, will answer / incorporate them tomorrow, and also rebase.

retronym commented:
@lrytz @adriaanm rebased and refined according to your review comments.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1208,10 +1200,38 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
+      val moduleNew = gen.newModule(module, moduleVar.tpe)
+      val body      = if (site.isTrait) moduleNew else {
+        if (moduleVar.enclClass == NoSymbol) gen.mkAssignAndReturn(moduleVar, moduleNew) // e.g. neg/t6666c.scala, will be an error later on.
+        else gen.mkAssignAndReturn(moduleVar, moduleNew) // will be wrapped in double checked lock idiom in lazyvals
```
the expressions in the `if` / `else` branches are identical

retronym replied:
Ha!

In an earlier version of this change, I synthesized all the lazy val ASTs right here, which hit a `NoSymbol` related error in some code that would have otherwise been flagged as erroneous in a later phase. So I only did that in the second branch of this `if`.

But now, since I defer the work until `lazyvals`, the second branch of the `if` became identical to the first :)

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1208,10 +1200,38 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
       transformTrees(newTrees map localTyper.typedPos(moduleDef.pos))
     }
+    def newInnerObject(site: Symbol, module: Symbol): List[Tree] = {
+      // Create the module var unless it is already in the module owner's scope.
```
is this comment correct? i don't see a test that would prevent something from being created.

retronym replied:
Nope, I should have removed the comment as part of resolving the merge conflicts after rebasing on your changes to this area.

lrytz commented on src/compiler/scala/tools/nsc/transform/LazyVals.scala:
```diff
@@ -100,11 +100,26 @@ abstract class LazyVals extends Transform with TypingTransformers with ast.TreeD
             sym.resetFlag((if (lazyUnit(sym)) 0 else LAZY) | ACCESSOR)
             (rhs1, sDef)
-          } else
+          } else if (sym.hasAllFlags(MODULE | METHOD) && !sym.owner.isTrait) {
```
Some comment would be helpful here. Maybe you could also extend a general doc comment somewhere (on `transform`?) to explain a bit more what's going on - eg what are "slowPath" and "fastPath"?

lrytz commented:
it looks very good - a bit like open heart surgery done by the best doctor in the field :) a bytecode diff maybe?

retronym commented:
Incorporated all your comments, @lrytz. I also realized I was still adding the module var to traits, I've removed that now.

Was:

``` scala
⚡ qscalac -Xprint:refchecks /Users/jason/code/scala/test/files/run/trait-defaults-modules.scala
[[syntax trees at end of                 refchecks]] // trait-defaults-modules.scala
package <empty> {
  abstract trait T1 extends scala.AnyRef {
    def a: Any
  };
  abstract trait T2 extends AnyRef with T1 {
    def /*T2*/$init$(): Unit = {
      ()
    };
    object a extends scala.AnyRef {
      def <init>(): T2.this.a.type = {
        a.super.<init>();
        ()
      }
    };
    @volatile <synthetic> private[this] var a$module: T2.this.a.type = _;
    <stable> def a: T2.this.a.type = new T2.this.a.type();
    object b extends scala.AnyRef {
      def <init>(): T2.this.b.type = {
        b.super.<init>();
        ()
      }
    };
    @volatile <synthetic> private[this] var b$module: T2.this.b.type = _;
    <stable> def b: T2.this.b.type = new T2.this.b.type();
    private object c extends scala.AnyRef {
      def <init>(): T2.this.c.type = {
        c.super.<init>();
        ()
      }
    };
    @volatile <synthetic> private[this] var c$module: T2.this.c.type = _;
    <stable> private def c: T2.this.c.type = new T2.this.c.type();
    def usec: Any = T2.this.c
  };
```

Now: 

``` scala
⚡ qscalac -Xprint:refchecks /Users/jason/code/scala/test/files/run/trait-defaults-modules.scala
[[syntax trees at end of                 refchecks]] // trait-defaults-modules.scala
package <empty> {
  abstract trait T1 extends scala.AnyRef {
    def a: Any
  };
  abstract trait T2 extends AnyRef with T1 {
    def /*T2*/$init$(): Unit = {
      ()
    };
    object a extends scala.AnyRef {
      def <init>(): T2.this.a.type = {
        a.super.<init>();
        ()
      }
    };
    <stable> def a: T2.this.a.type;
    object b extends scala.AnyRef {
      def <init>(): T2.this.b.type = {
        b.super.<init>();
        ()
      }
    };
    <stable> def b: T2.this.b.type;
    private object c extends scala.AnyRef {
      def <init>(): T2.this.c.type = {
        c.super.<init>();
        ()
      }
    };
    <stable> private def c: T2.this.c.type;
    def usec: Any = T2.this.c
  };
```

adriaanm commented on src/compiler/scala/tools/nsc/ast/TreeGen.scala:
```diff
@@ -242,11 +242,14 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
-      If(cond, Block(syncBody: _*), EmptyTree)) ::
-      stats: _*)
+  def mkSynchronizedCheck(attrThis: Tree, cond: Tree, syncBody: List[Tree], stats: List[Tree]): Tree = {
+    def blockOrStat(stats: List[Tree]): Tree = stats match {
```
Can we use the Block factory method here? https://github.com/scala/scala/blob/2.12.x/src/reflect/scala/reflect/internal/Trees.scala#L1181

retronym replied:
No, the intention here it to avoid a wrapper block altogether if `stats.size == 1`.

adriaanm replied:
Makes sense.

adriaanm commented on src/compiler/scala/tools/nsc/transform/LazyVals.scala:
```diff
@@ -54,14 +54,35 @@ abstract class LazyVals extends Transform with TypingTransformers with ast.TreeD
     private val lazyVals = perRunCaches.newMap[Symbol, Int]() withDefaultValue 0
 
     import symtab.Flags._
+    private def flattenThickets(stats: List[Tree]): List[Tree] = stats.flatMap(_ match {
```
I've also needed this, but the `stats` weren't necessarily `DefDef`s -- could a universal encoding of a thicket look like, e.g., `Block(stats, EmptyTree)`? Maybe we should call it a shrubbery, though.

retronym replied:
I've just tried this (outside of this PR) here: https://github.com/retronym/scala/compare/6fb41ed...retronym:topic/shrubbery?expand=1

I couldn't find too many spots that were ripe for retrofitting, unfortunately. Can you think of any others? If we can find more clients, the change will be more compelling.

/cc @densh

adriaanm replied:
I think there may be more opportunities in MethodSynthesis, but we can probably just make a mental note and come back to this.

adriaanm commented on src/compiler/scala/tools/nsc/transform/LazyVals.scala:
```diff
@@ -54,14 +54,35 @@ abstract class LazyVals extends Transform with TypingTransformers with ast.TreeD
      *  - remove ACCESSOR flags: accessors in traits are not statically implemented,
      *    but moved to the host class. local lazy values should be statically implemented.
+     *
+     * The general pattern is
```
"... wait for it!"

adriaanm commented on src/compiler/scala/tools/nsc/transform/LazyVals.scala:
```diff
@@ -100,11 +115,26 @@ abstract class LazyVals extends Transform with TypingTransformers with ast.TreeD
+            rhs match {
+              case b @ Block((assign @ Assign(moduleRef, _)) :: Nil, expr) =>
+                val cond = Apply(Select(moduleRef, Object_eq), List(Literal(Constant(null))))
+                val (fastPath, slowPath) = mkFastPathBody(sym.owner.enclClass, moduleRef.symbol, cond, transform(assign) :: Nil, Nil, transform(expr))
```
Looks like `cond` is shared between different subtrees.

adriaanm replied:
ping @retronym -- I think the sharing of `cond` is a problem, the rest of the PR LGTM

retronym replied:
Thanks for the reminder. On my TODO list for Monday.

retronym committed:
Avoid tree sharing in lazy val translation

The `cond` tree is used in two places, so make it a by-name parameter
to create distinct Trees.

Also, remove a dangling comment.

adriaanm commented:
LGTM!
